### ISSUE TITLE
Issue Re-checking files

### ISSUE THREAD

**Issue by miko3ed** [evilhero/mylar#1744]
Having an issue with an imported comic.
Downloaded it, ComicRack tagged it successfully, Mylar imported it with manual import, the directory for the comic was created and the imported comic was moved into the directory. The problem is that the comic list shows 0/1 in the have list, and when I open the details for it the issue is tagged as wanted. Running a manual Recheck Files finds the file but doesn't match it for some reason. Any help in figuring this one out would be appreciated.

Here is the log for the Recheck Files command:
> 2017-10-17 18:33:33 INFO    [FILE-RESCAN] I have physically found 0 issues, ignored 0 issues, snatched 0 issues, and accounted for 0 in an Archived state [ Total Issue Count: 0 / 1 ]
> 2017-10-17 18:33:33 DEBUG   Successfully changed ownership and permissions [99:100] / [0777 / 0666]
> 2017-10-17 18:33:33 DEBUG   [FILE-RESCAN] Ensuring permissions/ownership enforced for series: Free Comic Book Day 2016 (Civil War II)
> 2017-10-17 18:33:33 INFO    [FILE-RESCAN] Total files located: 0
> 2017-10-17 18:33:33 INFO    [FILE-RESCAN] Updated the status of 1 issues for Free Comic Book Day 2016 (Civil War II) (2016) that were not found.
> 2017-10-17 18:33:33 DEBUG   No issues with identical issue numbering were detected for this series
> 2017-10-17 18:33:33 INFO    FAILED FILES: [{'issue_number': '001', 'sub': None, 'issue_year': '2016', 'comiclocation': u'/comics/Free Comic Book Day 2016 (Civil War II) V2016 (2016)', 'scangroup': None, 'parse_status': 'failure', 'series_name': 'Free Comic Book Day 2016', 'comicfilename': 'Free Comic Book Day 2016 (Civil War II) V2016 #001 (2016).cbz', 'series_volume': 'v2016'}]
> 2017-10-17 18:33:33 INFO    [NO MATCH] Free Comic Book Day 2016 (Civil War II) V2016 #001 (2016).cbz [WATCHLIST:Free Comic Book Day 2016 (Civil War II)]
> 2017-10-17 18:33:33 DEBUG   Possible Alternate Names to match against (if necessary): ['127372873872871091383 abdkhjhskjhkjdhakajhf']
> 2017-10-17 18:33:33 INFO    Series_Name: Free Comic Book Day 2016 --- WatchComic: Free Comic Book Day 2016 (Civil War II)
> 2017-10-17 18:33:33 DEBUG   series title possibly: Free Comic Book Day 2016
> 2017-10-17 18:33:33 DEBUG   year 2016 is within series title. Ignoring as YEAR value
> 2017-10-17 18:33:33 DEBUG   year 2016 is outside of series title range. Accepting of year.
> 2017-10-17 18:33:33 DEBUG   Volume detected as : v2016
> 2017-10-17 18:33:33 DEBUG   new split: ['Free', 'Comic', 'Book', 'Day', '2016', '(Civil War II)', '001', 'V2016', '(2016)']
> 2017-10-17 18:33:33 DEBUG   Extra item(s) are present between the volume label and the issue number. Checking..
> 2017-10-17 18:33:33 DEBUG   issue verified as : 001
> 2017-10-17 18:33:33 DEBUG   issue_position: 7
> 2017-10-17 18:33:33 DEBUG   Already validated year, ignoring as possible issue number: 2016
> 2017-10-17 18:33:33 DEBUG   issue number :001
> 2017-10-17 18:33:33 DEBUG   possible_issuenumbers: [{'position': 4, 'validcountchk': False, 'number': '2016', 'mod_position': 20}, {'position': 7, 'validcountchk': False, 'number': '001', 'mod_position': 47}]
> 2017-10-17 18:33:33 DEBUG   highest_series_position: 4
> 2017-10-17 18:33:33 DEBUG   year verified as: 2016
> 2017-10-17 18:33:33 DEBUG   year verified as : 2016
> 2017-10-17 18:33:33 DEBUG   sctd: 2017
> 2017-10-17 18:33:33 DEBUG   year verified as: 2016
> 2017-10-17 18:33:33 DEBUG   year verified as : 2016
> 2017-10-17 18:33:33 DEBUG   sctd: 2017
> 2017-10-17 18:33:33 DEBUG   datecheck: [{'date': '2016', 'position': 4, 'mod_position': 20}, {'date': '2016', 'position': 8, 'mod_position': 51}]
> 2017-10-17 18:33:33 DEBUG   date: 2016
> 2017-10-17 18:33:33 DEBUG   checking date : 2016
> 2017-10-17 18:33:33 DEBUG   volume label detected as : Volume 2016 @ position: 6
> 2017-10-17 18:33:33 DEBUG   date: 2016
> 2017-10-17 18:33:33 DEBUG   checking date : 2016
> 2017-10-17 18:33:33 INFO    [FILENAME]: Free Comic Book Day 2016 (Civil War II) V2016 #001 (2016).cbz
> 2017-10-17 18:33:33 INFO    there are 1 files.
> 2017-10-17 18:33:33 INFO    [FILE-RESCAN] Now checking files for Free Comic Book Day 2016 (Civil War II) (2016) in /comics/Free Comic Book Day 2016 (Civil War II) V2016 (2016)
> 2017-10-17 18:33:30 DEBUG   If you read this message, debug logging is available
> 2017-10-17 18:33:30 INFO    Verbose toggled, set to True
> 2017-10-17 18:33:27 INFO    Web logs cleared

**Comment by evilhero** [comment]
Based on that log it looks like it's getting hung up in the either the '(2016)' or the '(Civil War II)', mainly because of the 2016 being duplicated and part of the series title having brackets with a year... I'll have to do some testing on this to see exactly where it's failing and what the fix would be.

**Comment by miko3ed** [comment]
If it helps
My Folder format is:
> $Series $VolumeY ($Year)

My File Format is: 
> $Series $VolumeY #$Issue ($Year)

**Comment by miko3ed** [comment]
It may also be worth mentioning that I have about 25 other series with this same issue. Here is the log snippet from another one:

> 2017-10-17 20:58:25 INFO    [FILE-RESCAN] I have physically found 0 issues, ignored 0 issues, snatched 0 issues, and accounted for 0 in an Archived state [ Total Issue Count: 0 / 1 ]
> 2017-10-17 20:58:25 DEBUG   Successfully changed ownership and permissions [99:100] / [0777 / 0666]
> 2017-10-17 20:58:25 DEBUG   [FILE-RESCAN] Ensuring permissions/ownership enforced for series: Batman/Deadman: Death and Glory
> 2017-10-17 20:58:25 INFO    [FILE-RESCAN] Total files located: 0
> 2017-10-17 20:58:25 INFO    [FILE-RESCAN] Updated the status of 1 issues for Batman/Deadman: Death and Glory (1996) that were not found.
> 2017-10-17 20:58:25 DEBUG   No issues with identical issue numbering were detected for this series
2017-10-17 20:58:25 INFO    FAILED FILES: [{'issue_number': '001', 'sub': None, 'issue_year': '1996', 'comiclocation': u'/comics/Batman-Deadman Death and Glory V1996 (1996)', 'scangroup': None, 'parse_status': 'failure', 'series_name': 'Batman-Deadman Death > and Glory', 'comicfilename': 'Batman-Deadman Death and Glory V1996 #001 (1996).cbz', 'series_volume': 'v1996'}]
> 2017-10-17 20:58:25 INFO    [NO MATCH] Batman-Deadman Death and Glory V1996 #001 (1996).cbz [WATCHLIST:Batman/Deadman: Death and Glory]
> 2017-10-17 20:58:25 DEBUG   Possible Alternate Names to match against (if necessary): ['127372873872871091383 abdkhjhskjhkjdhakajhf']
> 2017-10-17 20:58:25 INFO    Series_Name: Batman-Deadman Death and Glory --- WatchComic: Batman/Deadman: Death and Glory
> 2017-10-17 20:58:25 DEBUG   series title possibly: Batman-Deadman Death and Glory
> 2017-10-17 20:58:25 DEBUG   Volume detected as : v1996
> 2017-10-17 20:58:25 DEBUG   issue verified as : 001
> 2017-10-17 20:58:25 DEBUG   issue_position: 5
> 2017-10-17 20:58:25 DEBUG   possible_issuenumbers: [{'position': 5, 'validcountchk': False, 'number': '001', 'mod_position': 38}]
> 2017-10-17 20:58:25 DEBUG   highest_series_position: 6
> 2017-10-17 20:58:25 DEBUG   year verified as: 1996
> 2017-10-17 20:58:25 DEBUG   year verified as : 1996
> 2017-10-17 20:58:25 DEBUG   sctd: 2017
> 2017-10-17 20:58:25 DEBUG   datecheck: [{'date': '1996', 'position': 6, 'mod_position': 42}]
> 2017-10-17 20:58:25 DEBUG   date: 1996
> 2017-10-17 20:58:25 DEBUG   checking date : 1996
> 2017-10-17 20:58:25 DEBUG   volume label detected as : Volume 1996 @ position: 4
> 2017-10-17 20:58:25 INFO    [FILENAME]: Batman-Deadman Death and Glory V1996 #001 (1996).cbz
> 2017-10-17 20:58:25 INFO    there are 1 files.
> 2017-10-17 20:58:25 INFO    [FILE-RESCAN] Now checking files for Batman/Deadman: Death and Glory (1996) in /comics/Batman-Deadman Death and Glory V1996 (1996)
> 2017-10-17 20:58:18 INFO    Web logs cleared

**Comment by evilhero** [comment]
2 seperate problems it looks like - civil war II not matching is due to the brackets not being included in the parsing (cause typically series titles' don't have them)/and a combination of the year. The Death and Glory is due to the word 'and' being repeated multiple times in the series title, which is causing the parser problems as it strips out common words for better matching results. So it changes 'Batman/Deadman: Death and Glory' to 'BatmanDeadmanDeathGlory' for one, and to match against it changes it to 'BatmanDeadmeathGlory'.

I'm looking at these currently, so will see what I can come up with....

**Comment by miko3ed** [comment]
I see what you're saying,,, I took a little time and messed with the Alternate Search Names and Alternate File-naming options and was able to get my files to successfully recheck.

If it helps I named the FCBD one to: 
> Free Comic Book Day 2k15 Secret Wars
and the Batman one to: 
> Batman-Deadman-Death&Glory

**Comment by evilhero** [comment]
Well at least you got it to match - but this also exposed a subsequent problem with the way the parser handles common words that exist normally in the title, and ones that come to exist after stripping out various characters for matching consideration. 

I'm on this now and trying to figure out a way to handle the differences in a proper manner so as to not mess up existing series that are handled properly. Might still be a day or so before I can roll this fix out, but I'll get it.  

Thanks for reporting back what worked for you - it all helps in figuring out how best to lay out the parser problems.

**Comment by miko3ed** [comment]
Yeah, text parsing isn't easy, having the year in the title also seems to be understandably problematic. Re: common words like "and", something I've done in the past with data parsing to remove common words is to convert spaces to an uncommon character, like | and then then strip the common words, then strip out | from the string. Dunno if that helps, but maybe :)

**Comment by evilhero** [comment]
Yeah, it's very problematic - I mean since I redid the filechecker from scratch last year it's leaps & bounds above what it was prior to that, but it still has some problems that can occur due to the various naming conventions that people have with their comics. Mylar replaces the uncommon characters with a | already, but I never thought about doing that with spaces (since the problem is common words being formed after all the parsing has completed). 

I just tried replacing the spaces with a combo character to ensure it doesn't occur normally (I chose '%$' cause why not), and it looks like it's working now ;)  Thanks for the advice on the space conversion - it was doing it, but not with replacing with a different character so this opened things up now. 

I still have some testing to do against it, and against the standard issue checks - but right now, it looks like this might get fixed sooner rather than later.
